fsthttp/imageopto/imageopto.go

```diff
@@ -0,0 +1,964 @@
+	// transformation will fail with the error "Invalid transformation for requested image:
+	// Invalid crop, region out of bounds". This option will instead deliver the image as an
+	// intersection of the origin image and the specified cropped region. This avoids the error,
+	// but the resulting image may not be of the specified dimensions.
+	CropModeSafe = "safe"
+
+	// CropModeSmart enables content-aware algorithms to attempt to crop the image to the desired aspect ratio
+	// while intelligently focusing on the most important visual content, including the detection
+	// of faces.
+	CropModeSmart = "smart"
+)
+
+func (c CropMode) isSet() bool    { return c != "" }
+func (c CropMode) String() string { return string(c) }
+
+type areaState uint8
+
+const (
+	areaStateNone areaState = iota
+	areaStateAspectRatio
+	areaStateWidthHeight
+)
+
+type areaAspectRatio struct{ w, h uint32 }
+type areaWidthHeight struct{ w, h PixelsOrPercentage }
+
+// Area is an image area.
+type Area struct {
+	aspect      areaAspectRatio
+	widthHeight areaWidthHeight
+	state       areaState
+}
+
+// NewAreaAspectRatio specifies an area with the given width/height aspect ratio.
+func NewAreaAspectRatio(w, h uint32) Area {
+	return Area{
+		aspect: areaAspectRatio{w, h},
+		state:  areaStateAspectRatio,
+	}
+}
+
+// NewAreaWidthHeight specifies an area with precise number of pixels.
+func NewAreaWidthHeight(w, h PixelsOrPercentage) Area {
+	return Area{
+		widthHeight: areaWidthHeight{w, h},
+		state:       areaStateWidthHeight,
+	}
+}
+
+func (a *Area) validate() error {
+	if a.state == areaStateWidthHeight {
+		if a.widthHeight.w.isSet() {
+			if err := a.widthHeight.w.validate(); err != nil {
+				return err
+			}
+		}
+		if a.widthHeight.h.isSet() {
+			if err := a.widthHeight.h.validate(); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (area *Area) String() string {
+	switch area.state {
+	case areaStateNone:
+		return ""
+	case areaStateAspectRatio:
+		return fmt.Sprintf("%v:%v", area.aspect.w, area.aspect.h)
+	case areaStateWidthHeight:
+		return fmt.Sprintf("%v,%v", area.widthHeight.w.String(), area.widthHeight.h.String())
+	}
+	return "error"
+}
+
+// PointOrOffset specifies a position along an axis
+type PointOrOffset string
+
+func (p PointOrOffset) isSet() bool    { return p != "" }
+func (p PointOrOffset) String() string { return string(p) }
+
+// Position is a position within an image.
+//
+// Valid values:
+//
+//	x{x},y{y}
+//	offset-x{offset-x},offset-y{offset-y}
+//	x{x},offset-y{offset-y}
+//	offset-x{offset-x},y{y}
+type Position struct {
+	X PointOrOffset
+	Y PointOrOffset
+}
+
+func (p *Position) String() string {
+	if p.X.isSet() && p.Y.isSet() {
+		return p.X.String() + "," + p.Y.String()
+	}
+
+	if p.X.isSet() {
+		return p.X.String()
+	}
+
+	return p.Y.String()
+}
+
+var reXPosition = regexp.MustCompile("^(x[0-9]+p?)|(offset-x[0-9]+)$")
+var reYPosition = regexp.MustCompile("^(y[0-9]+p?)|(offset-y[0-9]+)$")
```

Comment:
There's a literal 'p' in the non-offset version?

Reply:
`x and y can be set as a value in pixels (e.g., 40 is 40 pixels) or as a percentage suffixed with p (e.g., 50p is 50%).`

Reply:
Hmm... I wonder if that needs to have a possible decimal for the percentage too.

Reply:
OK, cool, I just didn't see that in the example valid values comment.